src/boot/efi/util.c

```diff
@@ -760,12 +760,11 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         /* Point dp to the end node of the copied device path. */
         dp = (EFI_DEVICE_PATH *) ((uint8_t *) *ret_dp + dp_size - END_DEVICE_PATH_LENGTH);
 
-        /* Replace end node with file media device path. */
-        FILEPATH_DEVICE_PATH *file_dp = (FILEPATH_DEVICE_PATH *) dp;
-        file_dp->Header.Type = MEDIA_DEVICE_PATH;
-        file_dp->Header.SubType = MEDIA_FILEPATH_DP;
-        memcpy(&file_dp->PathName, file, file_size);
-        SetDevicePathNodeLength(&file_dp->Header, SIZE_OF_FILEPATH_DEVICE_PATH + file_size);
+        /* Replace end node with file media device path. Use memcpy in case dp is unaligned. */
+        dp->Type = MEDIA_DEVICE_PATH;
+        dp->SubType = MEDIA_FILEPATH_DP;
+        memcpy((uint8_t *) dp + offsetof(FILEPATH_DEVICE_PATH, PathName), file, file_size);
+        SetDevicePathNodeLength(dp, SIZE_OF_FILEPATH_DEVICE_PATH + file_size);
```

Comment:
not sure i follow here? if `dp` is aligned, then `file_dp` would be aligned too.

If things aren't aligned here, then you can't write to `dp->Type` either...

Reply:
`EFI_DEVICE_PATH` has alignment of 1, so it's always fine (It's all bytes in it. The length field is technically uint16_t, but it is deliberately split into 2 bytes). But that of `FILEPATH_DEVICE_PATH` is 2 (and `dp` at this point is supposed to point to that). Mind you, device paths are a packed structure and we generally cannot just assume that the size leading up to `dp` gives the right alignment. We also cannot just create padding as at least OVMF didn't like that for HD device paths.

Reply:
packed as in gcc attribute packed?

Reply:
> packed as in gcc attribute packed?

If only gnu-efi/edk2 would use that so we'd not have to care. One can dream… Some of their provided definitions even have tail padding, which at least OVMF doesn't like to see.

Quoting from the spec:
> Therefore all nodes are byte-packed data structures that may appear on any byte boundary. All code
references to device path notes [sic] must assume all fields are unaligned.

What's even more annoying that except for `EFI_DEVICE_PATH`, none of the device path structs are provide as struct definitions. They're all in (table-form) prose.

Reply:
hmm, yuck. This really sounds to me as if we should put together the structs in heap or stack and then copy them into place, to deal with the alignment mess.

Reply:
I fail to see the point in doing that here. You need to use memcpy regardles for FILEPATH_DEVICE_PATH, so I'm not going to do the same thing twice.